akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ShardingDirectReplication.scala

```diff
@@ -103,9 +103,11 @@ object ShardingDirectReplication {
             event.sequenceNumber)
           replicaShardingProxies.foreach {
             case (replica, proxy) =>
-              val envelopedEvent = ShardingEnvelope(event.persistenceId.id, event)
-              if (!selfReplica.contains(replica))
+              val newId = ReplicationId.fromString(event.persistenceId.id).withReplica(replica)
+              val envelopedEvent = ShardingEnvelope(newId.persistenceId.id, event)
+              if (!selfReplica.contains(replica)) {
                 proxy.asInstanceOf[ActorRef[ShardingEnvelope[PublishedEvent]]] ! envelopedEvent
```

Comment:
`unsafeUpcast` instead of instanceof ?

Reply:
This is too unsafe for that as it has a bounds check on the type param